[#7177] feat(core): Implement cache integration with relational

Abyss-lord opened the pull request:
### What changes were proposed in this pull request?

Change:
- Updated EntityCache interface: `get` no longer delegates to underlying store
- Enhanced RelationalEntityStore to support entity caching
- Revised multiple test cases to reflect caching behavior and validate correctness

Tasks:
- [X] Implement the methods related to the `SupportsEntityStoreCache` and `SupportsRelationEntityCache` interface in `CaffeineEntityCache`.
- [X] Implement cache integration with relational store and add basic unit tests. 
- [ ] Add unit tests for both index and cache.
- [ ] Use JCStress to perform multi-threaded testing on `CaffeineEntityCache` and the `CacheIndex`.

### Why are the changes needed?

Fix: #7177

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

local test.

xunliu commented on core/src/main/java/org/apache/gravitino/storage/relational/RelationalEntityStore.java:
```diff
@@ -54,12 +58,26 @@ public class RelationalEntityStore
     this.garbageCollector = new RelationalGarbageCollector(backend, config);
     this.garbageCollector.start();
+    // TODO USE SPI to load the cache
+    this.cache = new CaffeineEntityCache(config);
+    this.cacheEnabled = config.get(Configs.CACHE_ENABLED);
```
maybe we need modify these code?

```
this.cacheEnabled = config.get(Configs.CACHE_ENABLED);
    if (!cacheEnabled) {
      this.cache = new CaffeineEntityCache(config);
    }
```

Abyss-lord replied:
FIx

xunliu commented on core/src/main/java/org/apache/gravitino/storage/relational/RelationalEntityStore.java:
```diff
@@ -184,7 +281,26 @@ public List<TagEntity> associateTagsWithMetadataObject(
+            return entities.get();
+          }
+
+          List<E> EntitiesFromDb =
```
Maybe the var name`backendEntities` is better than `EntitiesFromDb`?

Abyss-lord replied:
Fix

xunliu commented on core/src/main/java/org/apache/gravitino/storage/relational/RelationalEntityStore.java:
```diff
@@ -83,46 +101,124 @@ private static RelationalBackend createRelationalEntityBackend(Config config) {
+
+    return cache.withCacheLock(
+        () -> {
+          List<E> entities = backend.list(namespace, entityType, false);
+          entities.forEach(cache::put);
+
+          return entities;
```
Can we list entities from cache?

xunliu commented on core/src/main/java/org/apache/gravitino/storage/relational/RelationalEntityStore.java:
```diff
@@ -83,46 +101,124 @@ private static RelationalBackend createRelationalEntityBackend(Config config) {
+
+    return cache.withCacheLock(
+        () -> {
+          List<E> entities = backend.list(namespace, entityType, allFields);
+          entities.forEach(cache::put);
+
+          return entities;
```
Can we list entities from cache?

Abyss-lord replied:
No, list operations must come from the underlying store to ensure data consistency and freshness. The cache is only used for get-by-id or name lookups.

xunliu commented on core/src/main/java/org/apache/gravitino/storage/relational/RelationalEntityStore.java:
```diff
@@ -83,46 +101,124 @@ private static RelationalBackend createRelationalEntityBackend(Config config) {
   @Override
   public boolean exists(NameIdentifier ident, Entity.EntityType entityType) throws IOException {
-    return backend.exists(ident, entityType);
+    if (!cacheEnabled) {
+      return backend.exists(ident, entityType);
+    }
+
+    return cache.withCacheLock(
+        () -> {
+          if (cache.contains(ident, entityType)) {
+            return true;
+          }
+
+          return backend.exists(ident, entityType);
+        });
```
I think maybe we need to adjust the code order?
We need to check the cache first, then the database

```
    if (cacheEnabled) {
      # check from cache
      # if exist then return xxx
    }

    return backend.exists(ident, entityType);
```

Abyss-lord replied:
Fix

yuqi1129 commented on core/src/main/java/org/apache/gravitino/storage/relational/RelationalEntityStore.java:
```diff
@@ -54,12 +58,28 @@ public class RelationalEntityStore
+   *
+   * @return {@code true} if cache is enable, otherwise {@code false}
+   */
+  public boolean cacheEnabled() {
```
This is unused. Is it for future use?

yuqi1129 commented on core/src/main/java/org/apache/gravitino/storage/relational/RelationalEntityStore.java:
```diff
@@ -83,47 +103,125 @@ private static RelationalBackend createRelationalEntityBackend(Config config) {
+    if (cacheEnabled) {
+      return cache.withCacheLock(
+          () -> {
+            List<E> entities = backend.list(namespace, entityType, allFields);
```
Will we store the entities every time `list` is called?  I believe in most cases, this is unnecessary except for the first one.

yuqi1129 commented on core/src/main/java/org/apache/gravitino/storage/relational/RelationalEntityStore.java:
```diff
@@ -83,47 +103,125 @@ private static RelationalBackend createRelationalEntityBackend(Config config) {
+              return true;
+            }
+
+            return backend.exists(ident, entityType);
```
I think this code can be moved out of the cache lock to reduce the race.

Abyss-lord replied:
Fix

yuqi1129 commented on core/src/main/java/org/apache/gravitino/storage/relational/RelationalEntityStore.java:
```diff
@@ -137,6 +235,7 @@ public <R, E extends Exception> R executeInTransaction(Executable<R, E> executab
   public void close() throws IOException {
     garbageCollector.close();
     backend.close();
+    cache.clear();
```
Close the cache first and then backend.

Abyss-lord replied:
Fix

yuqi1129 commented on core/src/test/java/org/apache/gravitino/utils/TestUtil.java:
```diff
@@ -0,0 +1,518 @@
+   * @return A list of the elements in the iterable.
+   * @param <T> The type of the elements in the iterable.
+   */
+  public static <T> List<T> toList(Iterable<T> iterable) {
```
I recall that they are similar tools that can provide this function. please refer to 
```
org.apache.commons.collections4.collect(Lists.newArrayList("a"), a -> a);
```

Abyss-lord replied:
Thanks for the suggestion! This method is actually not used in the current context, so I've decided to remove it to avoid confusion. Appreciate your input as always.

yuqi1129 commented on core/src/main/java/org/apache/gravitino/storage/relational/RelationalEntityStore.java:
```diff
@@ -196,6 +314,15 @@ public void insertRelation(
+      cache.withCacheLock(
+          () -> {
+            cache.invalidate(srcIdentifier, srcType, relType);
+            backend.insertRelation(
```
ditto

yuqi1129 replied:
This has not be resolved.

yuqi1129 commented on core/src/main/java/org/apache/gravitino/storage/relational/RelationalEntityStore.java:
```diff
@@ -83,47 +107,121 @@ private static RelationalBackend createRelationalEntityBackend(Config config) {
+        return cache.withCacheLock(
+            () -> {
+              cache.clear();
+              return backend.delete(ident, entityType, cascade);
```
This can also be moved out of the cache lock block.

Abyss-lord replied:
Based on offline discussion, confirmed that end-to-end consistency is not required in this scenario. That means the arrival order does not need to be strictly preserved across both the cache and the underlying database.

yuqi1129 commented:
> Hi @yuqi1129 @xunliu , I've completed the code updates and would appreciate your review of the PR when you have a moment. Here's a summary of the commits:
> 
> 1. Some backend logic moved out of the cache lock to reduce the race(`delete`、`insert` and `exists`).
> 2. Fix codes.

@xunliu
It's okay with me. Please go ahead.

xunliu commented:
@jerryshao Please help review this PR, Thanks.

jerryshao commented:
> @jerryshao Please help review this PR, Thanks.

I will review it soon.

jerryshao commented on core/src/main/java/org/apache/gravitino/storage/relational/RelationalEntityStore.java:
```diff
@@ -54,12 +60,30 @@ public class RelationalEntityStore
+    if (cacheEnabled) {
+      this.cache = new CaffeineEntityCache(config);
+      this.namespaceSet = Sets.newHashSet();
+    }
```
I think it is better to have a `NoOpsCache` if cache is not enabled, so you don't have to add lots of "if conditions" to check whether the cache is enabled or not.

Abyss-lord replied:
Fix

jerryshao commented on core/src/main/java/org/apache/gravitino/storage/relational/RelationalEntityStore.java:
```diff
@@ -54,12 +61,28 @@ public class RelationalEntityStore
     this.garbageCollector = new RelationalGarbageCollector(backend, config);
     this.garbageCollector.start();
+    this.cacheEnabled = config.get(Configs.CACHE_ENABLED);
+    this.namespaceSet = Sets.newHashSet();
```
What is the usage of `namespaceSet`?

Abyss-lord replied:
> What is the usage of `namespaceSet`?

> > Will we store the entities every time `list` is called? I believe in most cases, this is unnecessary except for the first one.
According to this opinion, it's used to cache the namespaces only during the first `list()` call. Subsequent calls will skip redundant additions.

jerryshao commented on core/src/main/java/org/apache/gravitino/storage/relational/RelationalEntityStore.java:
```diff
@@ -83,47 +106,91 @@ private static RelationalBackend createRelationalEntityBackend(Config config) {
   public <E extends Entity & HasIdentifier> List<E> list(
       Namespace namespace, Class<E> type, Entity.EntityType entityType) throws IOException {
-    return backend.list(namespace, entityType, false);
+    return cache.withCacheLock(
```
I found that you have the lock in the cache, also you will have a lock in the entity store. I would suggest using only one lock here, no need to use lock in the cache, or use a different lock in here? The use of one lock in two different classes easily leads to race conditions.

Abyss-lord replied:
In this case, both places are using the same reentrant lock instance. The intention is to ensure that access to the cache is thread-safe, whether it's used directly or through the EntityStore. So even when the cache is used independently, the lock is still necessary to maintain consistency.

Abyss-lord commented on core/src/main/java/org/apache/gravitino/storage/relational/RelationalEntityStore.java:
```diff
@@ -83,47 +106,91 @@ private static RelationalBackend createRelationalEntityBackend(Config config) {
+    return cache.withCacheLock(
+        () -> {
+          List<E> entities = backend.list(namespace, entityType, false);
+          if (namespaceSet.add(namespace)) {
+            entities.forEach(cache::put);
+          }
```
By populating the cache the first time a namespace is listed, we avoid repeated database lookups for entities under the same namespace. The namespaceSet.add(namespace) check ensures that this logic only runs once per namespace.

Abyss-lord replied:
@jerryshao The current implementation of `list()` no longer caches the returned entities into the cache layer as above.

Abyss-lord commented on core/src/main/java/org/apache/gravitino/storage/relational/RelationalEntityStore.java:
```diff
@@ -83,47 +106,91 @@ private static RelationalBackend createRelationalEntityBackend(Config config) {
+        () -> {
+          backend.insert(e, overwritten);
+
+          if (e.type() == Entity.EntityType.MODEL_VERSION) {
+            NameIdentifier modelIdent = ((ModelVersionEntity) e).modelIdentifier();
+            cache.invalidate(modelIdent, Entity.EntityType.MODEL);
+          }
```
> What is the purpose of this code?

Inserting a ModelVersion may update the corresponding Model entity (e.g., updating its latest version reference), so we need to invalidate the cache for the associated Model to ensure consistency.

Abyss-lord commented:
Hi @jerryshao , I've completed the code updates and would appreciate your review of the PR when you have a moment. Here's a summary of the commits:
1.  Updated the locking strategy:
    - In RelationalEntityStore, all operations except get and listEntitiesByRelation are no longer wrapped in withCacheLock.
    - In CaffeineEntityCache, only mutating operations (put, invalidate, clear, etc.) are protected by locks. Methods like contains and getIfPresent are left lock-free, since Caffeine is thread-safe for read access.
2. Introduced SupportsEntityStoreCache interface. It's designed specifically to handle cache invalidation scenarios triggered by inserting entities like ModelVersion that may affect related entries such as Model.
3.  Removed the usage of namespaceSet

jerryshao commented on core/src/main/java/org/apache/gravitino/cache/BaseEntityCache.java:
```diff
@@ -133,7 +76,7 @@ protected static void validateEntityHasIdentifier(Entity entity) {
    */
   @SuppressWarnings("unchecked")
-  protected static <E extends Entity & HasIdentifier> List<E> convertEntity(List<Entity> entities) {
+  public static <E extends Entity & HasIdentifier> List<E> convertEntity(List<Entity> entities) {
```
Please change to `convertEntities`.

Abyss-lord replied:
Fix

jerryshao commented on core/src/main/java/org/apache/gravitino/cache/CaffeineEntityCache.java:
```diff
@@ -306,12 +207,23 @@ public <E extends Entity & HasIdentifier> void put(E entity) {
+      NameIdentifier modelIdent = ((ModelVersionEntity) entity).modelIdentifier();
+      invalidate(modelIdent, Entity.EntityType.MODEL);
+    }
+  }
```
Please add comment to the code to explain why do you need this.

Abyss-lord replied:
Fix

jerryshao commented on core/src/main/java/org/apache/gravitino/cache/NoOpsCache.java:
```diff
@@ -0,0 +1,134 @@
+
+  /** {@inheritDoc} */
+  @Override
+  public <E extends Exception> void withCacheLock(ThrowingRunnable<E> action) throws E {
+    action.run();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public <T, E extends Exception> T withCacheLock(ThrowingSupplier<T, E> action) throws E {
+    return action.get();
+  }
```
If user they want to use cache lock to guarantee the atomicity, I think we still need a lock here even for `NoOpsCache`, what do you think?

jerryshao commented on core/src/main/java/org/apache/gravitino/storage/relational/RelationalEntityStore.java:
```diff
@@ -54,12 +58,26 @@ public class RelationalEntityStore
           Configs.DEFAULT_ENTITY_RELATIONAL_STORE, JDBCBackend.class.getCanonicalName());
   private RelationalBackend backend;
   private RelationalGarbageCollector garbageCollector;
+  private EntityCache cache;
+  private boolean cacheEnabled;
+
+  /**
+   * Return whether the cache is enabled or not.
+   *
+   * @return {@code true} if cache is enable, otherwise {@code false}
+   */
+  public boolean cacheEnabled() {
```
Do you still need this method?

Abyss-lord replied:
No,  I’ve already removed this method.

jerryshao commented on core/src/main/java/org/apache/gravitino/storage/relational/RelationalEntityStore.java:
```diff
@@ -184,7 +217,20 @@ public List<TagEntity> associateTagsWithMetadataObject(
+          List<E> backendEntities =
+              backend.listEntitiesByRelation(relType, nameIdentifier, identType, allFields);
+
+          cache.put(nameIdentifier, identType, relType, backendEntities);
```
My concern is that the current cache may not be balanced. For some keys the value list may be large (like here). So the size  of cache may not be accurate, and potentially lead to OOM. For example, if the relation list is very large.

jerryshao replied:
Also you only cache the list relations, not the list operation, what is your consideration?

jerryshao replied:
If we want to manage the cache very well, to avoid the problem I mentioned above, my feeling is that current kv map is not enough. You need to design your own data structure, like linked hashmap, or something else.

Abyss-lord replied:
> My concern is that the current cache may not be balanced. For some keys the value list may be large (like here). So the size of cache may not be accurate, and potentially lead to OOM. For example, if the relation list is very large.

Thanks for the suggestion! I've added weight-based control in the `weigh(...)` method. The total weight is now calculated per entity using `calculateWeight(type)`, and will not cache the entry if the total exceeds the maxWeight threshold. This helps prevent oversized entries from polluting the cache.
